File: struct_simd/only_simd/bench.jl
using Statistics
using Printf



function compute_simd(x::Array{Float64, 1}, y::Array{Float64, 1},
                 grads::Array{Float64, 3}, out::Array{Float64, 1})

    det = -(x[2] - x[3])*y[1] + (x[1] - x[3])*y[2] - (x[1] - x[2])*y[3]
    dv = 1.0 / (6.0 * det)
    ii = 1
    for i=1:6
        for j=1:i
            s = 0.0
            @simd for k=1:3
                s += grads[1, k, i] * grads[1, k, j] + grads[2, k, i] * grads[2, k, j]
            end
            out[ii] = dv * s
            ii += 1
        end
    end

    return out
end

function compute(x::Array{Float64, 1}, y::Array{Float64, 1},
                 grads::Array{Float64, 3}, out::Array{Float64, 1})

    det = -(x[2] - x[3])*y[1] + (x[1] - x[3])*y[2] - (x[1] - x[2])*y[3]
    dv = 1.0 / (6.0 * det)
    ii = 1
    for i=1:6
        for j=1:i
            s = 0.0
            for k=1:3
                s += grads[1, k, i] * grads[1, k, j] + grads[2, k, i] * grads[2, k, j]
            end
            out[ii] = dv * s
            ii += 1
        end
    end

    return out
end


const x = Float64[0., 1., 0.]
const y = Float64[0., 0., 1.]

a11 = -y[1] + y[3]
a12 =  y[1] - y[2]
a21 =  x[1] - x[3]
a22 = -x[1] + x[2]

gq = reshape(
    [-1.0,-1.0,1.0,1.0,-1.0,-1.0,1.0,0.0,1.0,0.0,-1.0,0.0,0.0,
     -1.0,0.0,1.0,0.0,1.0,0.0,-2.0,-2.0,-2.0,2.0,0.0,0.0,2.0,2.0,
     2.0,2.0,0.0,0.0,2.0,-2.0,-2.0,-2.0,0], 2, 3, 6)

grads = zeros(2, 3, 6)

for f=1:6
    for p=1:3
        grads[1, p, f] = a11 * gq[1, p, f] + a12 * gq[2, p, f]
        grads[2, p, f] = a21 * gq[1, p, f] + a22 * gq[2, p, f]
    end
end


result = zeros(21)

nb_runs = 1000

times = zeros(nb_runs)

for irun in 1:nb_runs
    times[irun] = @elapsed compute(x, y, grads, result)
end

time_without_simd = median(times)

@printf("Julia without simd %.3f µs\n", time_without_simd * 1e6)

for irun in 1:nb_runs
    times[irun] = @elapsed compute_simd(x, y, grads, result)
end

time_with_simd = median(times)

@printf("Julia with simd    %.3f µs\n", time_with_simd * 1e6)


@printf("ratio (without simd)/(with simd): %.2f\n", time_without_simd / time_with_simd)

File: struct_simd/only_simd/bench.py
import numpy as np

from transonic import boost

A1 = "float[]"


@boost
def compute(x: A1, y: A1, grads: "float[:, :, :]", out: A1):
    det = -(x[1] - x[2]) * y[0] + (x[0] - x[2]) * y[1] - (x[0] - x[1]) * y[2]
    dv = 1.0 / (6.0 * det)
    ii = 0
    for i in range(6):
        for j in range(i + 1):
            s = 0.0
            for k in range(3):
                s += (
                    grads[i, k, 0] * grads[j, k, 0]
                    + grads[i, k, 1] * grads[j, k, 1]
                )
            out[ii] = dv * s
            ii += 1


if __name__ == "__main__":

    from functools import partial
    from transonic.util import timeit

    x = np.array([0.0, 1.0, 0.0])
    y = np.array([0.0, 0.0, 1.0])

    grads = np.empty((6, 3, 2))

    # fmt: off
    gq = np.array(
        [-1.0, -1.0, 1.0, 1.0, -1.0, -1.0, 1.0, 0.0, 1.0, 0.0, -1.0,
         0.0, 0.0, -1.0, 0.0, 1.0, 0.0, 1.0, 0.0, -2.0, -2.0, -2.0,
         2.0, 0.0, 0.0, 2.0, 2.0, 2.0, 2.0, 0.0, 0.0, 2.0, -2.0, -2.0,
         -2.0, 0.0]
    )
    # fmt: on

    gq = gq.reshape(grads.shape)
    a11 = -y[0] + y[2]
    a12 = y[0] - y[1]
    a21 = x[0] - x[2]
    a22 = -x[0] + x[1]

    for f in range(6):
        for p in range(3):
            grads[f, p, 0] = a11 * gq[f, p, 0] + a12 * gq[f, p, 1]
            grads[f, p, 1] = a21 * gq[f, p, 0] + a22 * gq[f, p, 1]

    result = np.zeros(21)

    compute(x, y, grads, result)

    timeit = partial(timeit, globals=locals(), total_duration=8)
    time = timeit("compute(x, y, grads, result)") * 1e6
    print(f"Pythran: {time:.2f} µs")

    from pathlib import Path
    import sys

    path_tmp_julia = Path("tmp_result_julia.txt")
    if not path_tmp_julia.exists():
        sys.exit()

    with open(path_tmp_julia) as file:
        txt = file.read()

    time_julia = float(txt.split("\n")[1].split(" ")[-2])

    print(f"Julia:   {time_julia:.2f} µs\nratio Pythran/Julia: {time/time_julia:.2f}")
